docs/rfcs/0011-script-run-stage.md

```diff
@@ -0,0 +1,208 @@
+
+# Detailed design
+
+## feature
+
+1. execute any commands in their pipeline.
+
+```yaml
+apiVersion: pipecd.dev/v1beta1
+kind: LambdaApp
+spec:
+  encryptedSecrets:
+    password: encrypted-secrets
+  pipeline:
+    stages:
+      - name: SCRIPT_RUN
+        with:
+            env:
+              AWS_PROFILE: default
+            runs:
+              - "echo {{ .encryptedSecrets.password }} | sudo -S su"
+              - "sam build"
+              - "sam deploy -g --profile $AWS_PROFILE"
+```
+
+2. combine with other stage
+
+Compared to CUSTOM_SYNC, this stage can be combined with other stage.
+For example, 
+
+```yaml
+apiVersion: pipecd.dev/v1beta1
+kind: KubernetesApp
+spec:
+  pipeline:
+    stages:
+      - name: K8S_CANARY_ROLLOUT
+        with:
+          replicas: 10%
+      - name: WAIT_APPROVAL
+        with:
+          timeout: 30m
+      - name: K8S_PRIMARY_ROLLOUT
+      - name: K8S_CANARY_CLEAN
+      - name: SCRIPT_RUN
+        with:
+            env:
+              SLACK_WEBHOOK_URL: ""
+            runs:
+              - "curl -X POST -H 'Content-type: application/json' --data '{"text":"successfully deployed!!"}' $SLACK_WEBHOOK_URL"
+```
+
+## when to rollback
+
+Users can define commands to execute with `onRollback` when rolling back.
+If `onRollback` is not set, nothing to execute when rolling back.
+
+```yaml
+apiVersion: pipecd.dev/v1beta1
+kind: KubernetesApp
+spec:
+  pipeline:
+    stages:
+      - name: SCRIPT_RUN
+        with:
+          env:
+            SLACK_WEBHOOK_URL: ""
+          runs:
+            - "curl -X POST -H 'Content-type: application/json' --data '{"text":"successfully deployed!!"}' $SLACK_WEBHOOK_URL"
+          onRollback:
+            - "curl -X POST -H 'Content-type: application/json' --data '{"text":"failed to deploy: rollback"}' $SLACK_WEBHOOK_URL"
+```
+
+**SCRIPT_SYNC stage also rollbacks** when the deployment status is `DeploymentStatus_DEPLOYMENT_CANCELLED` or `DeploymentStatus_DEPLOYMENT_FAILURE` even though other rollback stage is also executed.
+
+For example, here is a deploy pipeline combined with other k8s stages.
+The result status of the pipeline is FAIL or CANCELED, piped rollbacks the stages `K8S_CANARY_ROLLOUT`, `K8S_PRIMARY_ROLLOUT`, and `SCRIPT_RUN`.
+
+```yaml
+apiVersion: pipecd.dev/v1beta1
+kind: KubernetesApp
+spec:
+  pipeline:
+    stages:
+      - name: K8S_CANARY_ROLLOUT
+        with:
+          replicas: 10%
+      - name: WAIT_APPROVAL
+        with:
+          timeout: 30m
+      - name: K8S_PRIMARY_ROLLOUT
+      - name: K8S_CANARY_CLEAN
+      - name: SCRIPT_RUN
+        with:
+          env:
+            SLACK_WEBHOOK_URL: ""
+          runs:
+            - "curl -X POST -H 'Content-type: application/json' --data '{"text":"successfully deployed!!"}' $SLACK_WEBHOOK_URL"
+          onRollback:
+            - "curl -X POST -H 'Content-type: application/json' --data '{"text":"failed to deploy: rollback"}' $SLACK_WEBHOOK_URL"
+```
+
```

Comment:
Some other questions I think we should define before starting:
- Do we support multi script_run stages combine
- How the user-defined `onRollback` be run in case of multi script_run stages combine
- How the user-defined `onRollback` be run in combination with the `XXX_ROLLBACK` stage? I mean, if k8s app is rollbacked and the pipeline contains script_run, how do we perform the rollback process? Should it be `XXX_ROLLBACK` first then `SCRIPT_RUN_ROLLBACK`?

Reply:
Thank you for your questions!
I will write about these things on rfc, but in advance, I will write my opinions on this .

> Do we support multi script_run stages combine
Yes, I think so. 

> How the user-defined onRollback be run in case of multi script_run stages combine
Multi script_run stages can be executed in the order these are set.

> How the user-defined onRollback be run in combination with the XXX_ROLLBACK stage? I mean, if k8s app is rollbacked and the pipeline contains script_run, how do we perform the rollback process? Should it be XXX_ROLLBACK first then SCRIPT_RUN_ROLLBACK?

I don't think about it. I will try to check the implementation and propose it 🙇

Reply:
> Multi script_run stages can be executed in the order these are set.

Either with their rollback script?

Reply:
I think it will be 2. (But I don't know yet it can be realized technically 🙇 )
xxx_canary -> script_run -> script_run -> rollback -> script_rollback -> script_rollback